crypto/bn/bn_lib.c

```diff
@@ -414,24 +415,30 @@ BIGNUM *BN_bin2bn(const unsigned char *s, int len, BIGNUM *ret)
 /* ignore negative */
 static int bn2binpad(const BIGNUM *a, unsigned char *to, int tolen)
 {
-    int i;
+    int i, j, top;
     BN_ULONG l;
 
-    bn_check_top(a);
     i = BN_num_bytes(a);
     if (tolen == -1)
         tolen = i;
     else if (tolen < i)
         return -1;
-    /* Add leading zeroes if necessary */
-    if (tolen > i) {
-        memset(to, 0, tolen - i);
-        to += tolen - i;
+
+    if (i == 0) {
+        OPENSSL_cleanse(to, tolen);
+        return tolen;
     }
-    while (i--) {
+
+    top = a->top * BN_BYTES;
+    for (i = 0, j = tolen; j > 0; i++) {
+        unsigned int mask;
+
+        mask = constant_time_lt(i, top);
+        i -= 1 & ~mask; /* stay on top limb */
         l = a->d[i / BN_BYTES];
```

Comment:
Isn't asan going to complain about this?

Reply:
No, it's top a->d[a->top-1] limb, most significant byte. But I can add [extended tests]...

Reply:
Oh, you stay on the top one. I'm wondering how much this leaks.

Reply:
Commit message discusses it: "leak is possible only if attacker can observe memory access pattern with limb granularity." Note that customarily attacker can observe with cache-line granularity. Even though it was shown that even limb granularity is possible on some processor, it could be measured only across a lot of accesses. This is rare operation in context.

Just in case for reference, this is actually small problem in the context, and there are more things pending, more essential...

Reply:
For reference, primary purpose was to get rid of padding memset in original implementation. Leak there was rather that depending on length it could take different code paths and attacker could observe that. It even might be possible without any active interference. If code that executed earlier is large enough to evict relevant fragments, and different memset code paths are of different sizes, timing would be able to just tell you that. Note that it's not about variation in data access (complete vector will have to be brought to cache in either case, so you won't be able tell the difference if you only observe cache in post-math), but rather access to code. One could replace it with memset that clears whole vector to eliminate this ambiguity and then write the data, but I figured that suggested approach lifts more questions.

Reply:
> it's not about variation in data access

This might be viewed controversial. It's not like there are no variations, but rather that variations are [relatively] small and it's not obvious that attacker would be able to use them *reliably*. Reliable variations stem rather from distinctly different *cache* access patterns, like disjoint cache lines (which would require active interference from attacker) or different amount of cache lines (can work for relatively passive attacker). In this case data access is cache-access neutral in sense that same cache lines are accessed irregardless input. Yes, this implies that input BN doesn't actually vary in length very much. And it doesn't in RSA case. However, if we were to close even that variation, then it would be possible by ensuring (or assuming) that a->dmax is large enough and perform even i<dmax comparison in addition to i<top. [But there is variation of similar nature in BN_num_bytes. It's different story and we'll get to that. And when we do, this might be revisited too.]